sky/adaptors/aws.py

```diff
@@ -27,6 +40,10 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+# Creating the session object is not thread-safe for boto3,
+# so we add a reentrant lock to synchronize the session creation.
+# Reference: https://github.com/boto/boto3/issues/1592
+@_synchronized
```

Comment:
Shall we test the locking approach's overheads for `status -r`, vs. making the session a ThreadLocal?

Reply:
Just updated the PR description with the overhead. It seems cannot make the session a ThreadLocal as the creation of sessions should be locked for multiple threads, but the same session can be used by multiple threads. Making the session a ThreadLocal will not help with the parallel creation of the session in multiple threads.

Reply:
Can we do something like
```python
thread_local = threading.local()
thread_local.session = None
...

@import_package
def session():
  if thread_local.session is None:
    thread_local.session = boto3.session.Session()
  return thread_local.session
```
Then each thread will use its own Session object. This may avoid locking overheads a bit. Note here `@functools.lru_cache()` is not needed which may have caused some weird issues?

Reply:
After disccused offline, we reached the consensus that the `threading.Local` is not a sufficient solution for the problem.